Road extension

raunakbh92 opened the pull request:
The `road_extension` branch provides
- Extensions to the interaction merge roadway to avoid vehicle reversals due to road ending
- Cooperative IDM driver model to do merging
- Merge overlay functionality

Commit:
road with downstream extensions built

Commit:
extended road building function

Commit:
Reel as dependency, joinpath to avoid hardcoding read track and roadway, avoid providing default roadway and traj kwargs, add tests

Commit:
barrier vehicle experiment

Commit:
wip: debug merging collision, relies on merging branch in AutomotiveDrivingModels.jl

Commit:
curvept and lane overlays, jumpy vehicle test

Commit:
wip:adapting cooperative IDM for the Interaction scenario

Commit:
wip: cooperative idm to handle merge scenario

Commit:
merging works using cooperative idm. Overlay added for merge ray tracing

MaximeBouton commented on src/cooperative_IDM.jl:
```diff
@@ -13,18 +13,12 @@ the car follows the IntelligentDriverModel.
     - `c::Float64 = 0.0` the cooperation level
-    - `fov::Float64 = 20.0` [m] A field of view, the merging vehicle is not considered if it is further than `fov`
 """
 @with_kw mutable struct CooperativeIDM <: DriverModel{LaneFollowingAccel}
```
Should this belon to the main ADM repo?

raunakbh92 replied:
I have significantly modified it from your version because there wasn't a clear main lane, merge lane distinction, and also there were multiple vehicles on what you called merge lane as opposed to single vehicle. Also, as it is now, it is specific to this roadway by hardcoding the lanes where merging takes place. So, I am wondering how to stress test it enough to make it a part of ADM (or AutomotiveSimulator.jl after name change)

MaximeBouton commented on src/overlays.jl:
```diff
@@ -80,3 +80,25 @@ function AutoViz.render!(rendermodel::RenderModel, overlay::LaneOverlay, scene::
+- Draw line joining vehicle and its associated merge vehicle
+- Inspired by MergeNeighborsOverlay from AutonomousMerging.jl
+"""
+@with_kw mutable struct MergeOverlay <: SceneOverlay
```
See the new AutoViz tutorial for making overlays

raunakbh92 committed:
Tweak roadway render! method from AutoViz to enable merge lane dashed boundary to look more sensible i.e. not go till very end as previously

MaximeBouton commented on src/roadway_building.jl:
````diff
@@ -375,3 +375,181 @@ function show_lane_overlays(road_ext::Roadway,traj_ext)
+render(road_ext)
+```
+"""
+function AutoViz.render!(rendermodel::RenderModel, roadway::Roadway;
````
I don't think this is the proper way to do this! 
It is a lot of repeated code and if we ever update the functions in AutoViz the changes will not follow here. 

Two suggestions: 
1- fix the bug in AutoViz by submitting a PR there that address your problem 
2- create a new roadway object with its own `add_renderable!`  function and try to limit repeated code as much as possible 

1 is definitely the preferred option here!

MaximeBouton commented on src/colorscheme.jl:
```diff
@@ -0,0 +1,81 @@
+const MONOKAY = Dict(
```
Avoid copying the stuff from AutoViz in here it is going to be a nightmare to maintain!
There might be a way to support what you want within AutoViz

MaximeBouton commented on src/AutomotiveInteraction.jl:
```diff
@@ -36,7 +36,8 @@ export
     make_discont_roadway_straight,
-    make_discont_roadway_jagged
+    make_discont_roadway_jagged,
+    render!
```
Look at the docs for AutoViz v0.8, `render!`  does not exist anymore.

raunakbh92 committed:
[WIP]: Compatibility with AutomotiveSimulator.jl and AutomotiveVisualization.jl

TODOS:
- Roadway wrapper with dedicated render function for dashed lane boundaries for merge lanes to be sensible
- Had to revive some of the functions from Records.jl/src/ListRecords.jl to do the vehicle track reading. See if these can be modernized

MaximeBouton commented on src/veh_track_reading.jl:
```diff
@@ -107,6 +107,49 @@ function car_df_index(trajdata::INTERACTIONTrajdata, carid::Int, frame::Int)
+    id::I
+end
+
+mutable struct ListRecord{S,D,I} # State, Definition, Identification
```
why not using arrays?

raunakbh92 committed:
Removed dependency on Records.jl functions for vehicle track reading

Also modified the video_trajdata_replay function to work with this modernized trajdata

TODOS:
- Figure out how to make IDOverlay(scene=trajdata[1]) work without complaining about Scene{Entity}
- merge lane specific roadway rendering

raunakbh92 committed:
Enable IDOverlay on replay trajdata

Was erroring out in previous commit with following error message:
julia> render([road_ext,trajdata[1],IDOverlay(scene=trajdata[1])])
ERROR: MethodError: no method matching IDOverlay(::Scene{Entity}, ::RGB{FixedPointNumbers.Normed{UInt8,8}}, ::Int64, ::Float64, ::Float64)
Closest candidates are:
  IDOverlay(::F, ::Any, ::Any, ::Any, ::Any) where F<:(Scene{Entity{S,D,I}} where I where D where S) at /home/raunakbh/.julia/packages/Parameters/l76EM/src/Parameters.jl:499
Stacktrace:

raunakbh92 committed:
Render avoid lane boundaries for merge lanes being weird. This time adhering to AutomotiveVisualization

- Wrapper around roadway with dedicated AutomotiveVisualization.add_renderable! implemented
- Caveat: Still copies a bunch of code from the roadway rendering functionality in AutomotiveVisualization.jl/src/roadways.jl

raunakbh92 committed:
Upgrade simulation and associated visualization functionality to AutomotiveSimulator and AutomotiveVisualization

Also, updated readme with new replay gif, and docs to point to dev instead of latest. Hopefully docs will show up from badge

raunakbh92 committed:
Merge branch 'master' into road_extension

MaximeBouton commented on src/visualization.jl:
```diff
@@ -35,22 +38,22 @@ end
     frames = Frames(MIME("image/png"),fps = 10)
-    
+    mr = MergingRoadway(roadway) # Wrapper for specialized render for merging lanes
+    mp = VecE2(1064.5227, 959.1559)
```
maybe store this in a constant in the main scope of the package instead of hardcoding it in the function.

MaximeBouton replied:
especially since you are reusing it at several places

raunakbh92 replied:
Removed the set_render_mode("fancy) and instead used `FancyCar` within the creation of `renderables`.

MaximeBouton commented on src/driving_simulation.jl:
```diff
@@ -274,14 +275,14 @@ test_barrier_vehicle(id_list=[20,29,19,28,6,8,25,2,10,7,18,12,100],roadway=road_
     veh20_state = veh_20.state
     deleteat!(scene,findfirst(20,scene)) # Remove veh 20 from scene
     barrier_state = VehicleState(veh20_state.posG,veh20_state.posF,0.)
     barrier_id = 100
```
use a constant

MaximeBouton commented on src/driving_simulation.jl:
```diff
@@ -274,14 +275,14 @@ test_barrier_vehicle(id_list=[20,29,19,28,6,8,25,2,10,7,18,12,100],roadway=road_
     keep_vehicle_subset!(scene,id_list)
     veh_20 = scene[findfirst(20,scene)]
     veh20_state = veh_20.state
     deleteat!(scene,findfirst(20,scene)) # Remove veh 20 from scene
```
maybe you can store the index and only call `findfirst` once

raunakbh92 committed:
overlay ground truth over model driven

raunakbh92 committed:
adding docs
